### ISSUE TITLE
How to use matcher to see if set contains or map containsKey or map (for key) has value?

### ISSUE THREAD

**Issue by cropredyHelix** [financialforcedev/fflib-apex-mocks#51]
Looking through the examples in `fflib_Match.cls` and `fflib_ApexMocksTest.cls` I'm not seeing any way to 

- see if a set contains an entry
- see if a map contains a key
- see if a map for key `k`, has value `v`

My example is testing a VF controller that calls a `FooService.doBar(..)`  Hence, mocking a service `FooServiceImpl`, method `doBar(map<ID,Foo.CustomType>)`

I tried this (match on entire map)

```
// preamble to set up mocks and inject not shown ; it works fine
((FooServiceImpl) mocks.verify(mockFooSvc,1))
  .doBar(new map<ID,Foo.CustomType> {someId => new Foo.CustomType(arg1, arg2)});
```

but even though `doBar` is called with the map `someId => new Foo.CustomType(arg1, arg2)` (verified using debug), the `verify` method in ApexMocks comes back with

`Expected : 1, Actual: 0 -- Wanted but not invoked: FooServiceImpl__sfdc_ApexStub.doBar(Map<Id,Foo.CustomType>)`

So, bottom line - would be nice to see how to use matchers on sets (specific entry) and maps (entire map, hasKey, hasValueForKey, etc.) - especially without having to resort to custom matchers.

**Comment by cropredyHelix** [comment]
After some inspired searching on [frup42 blog](https://superdupercode.blogspot.com/2016/06/apexmocks-capture-support.html), I now see how to use `fflib_ArgumentCaptor` to solve the above on verifying maps.   Still seems as if there should be enhancement possibilities for more matchers on `fflib_Match` to deal with sets.

**Comment by dfruddffdc** [comment]
Another Apex quirk... This compiles fine:
```java
List<Account> records = new List<Account>{};
List<Object> castUp = (List<Object>)records;
List<Account> castDown = (List<Account>)castUp;
```
But this doesn't compile, because: **Incompatible types since an instance of Set<Account> is never an instance of Set<Object>**
```java
Set<Account> records = new Set<Account>{};
Set<Object> castUp = (Set<Object>)records;
Set<Account> castDown = (Set<Account>)castUp;
```

Likewise with Maps. So we can't write generic matchers for Maps and Sets, we have to have a specific matcher for the specific Typed Maps/Sets. This would be a lot of matchers, but not particularly useful for the majority of ApexMocks consumers.

So when the requirement comes up, we just let the consumer use captors or custom matchers knowing the specific Types in question in their test.
